### DISCUSSION TITLE
Feature Request: The ability to provide columns in conjunction with auto_detect=true

### DISCUSSION THREAD

**Discussion by skinkie** [duckdb/duckdb#9538] [Ideas]
When more complex CSV files are being process, some standardisation has taken place. I can fill in `columns={...}` with the columns that I know that will exists. What can't be done is stating: _for the headers provided, in any order, this will be the matching type, but if you encounter any other column do auto_detect=true_.

**Comment by shreeve** [comment]
This would be very helpful. For example, I commonly don't mind that all fields are loaded as `varchar`, but I may want to force a field or two to be loaded as `decimal(9,2)` (like a currency). It would be great to provide this kind of _hint_ without forcing an order or requiring all fields to be defined.

**Reply by skinkie** [reply]
And I guess this would be pushing it... but if the column is described, and not part of the data, I would like to have the column added.

**Comment by carlopi** [comment]
Related: @pdet work on `sniff_csv` https://github.com/duckdb/duckdb/pull/9522.

When that makes into mainline DuckDB it would become possible to build more complex functionality around CSV columns, like have sniff_csv returns some types, override them with others, and then do the reading.

**Comment by FrancoisLepoutre** [comment]
> "What can't be done is stating: for the headers provided, in any order, this will be the matching type, but if you encounter any other column do auto_detect=true"

> When that makes into mainline DuckDB it would become possible to build more complex functionality around CSV columns, like have sniff_csv returns some types, override them with others, and then do the reading

Hi everyone,

Just a lurker into this conversation and the great work and blogs provided on this interesting (and time-consuming) issue: csv feeding! 

**I find this proposal and this thread quite interesting.** In some way, I would push for the idea that users could provide a full, possibly quite large and possibly unlimited list of "column typing constraints" to this great feeding beast. In case the name information provided, the type would handled i-e essentially casted as requested

The current syntax for type,dtypes is ok. The idea one could provide a "(potentially very) large set of mapping constraints" as some sort of general pattern whatever the feeding material would possibly fit many!

**Naming is an essential time-consuming activity for a significant number of db specialists.** Specifically when it comes to large datasets, I mean "wide ones", it requires verbal skills and, at times, lexical and orthographical ones! 

An ability to feed long dtypes-types schemes, possibly alphabetically ordered ones of course, that could apply cross table boundaries could possibly save some time and avoid quite a number of generic type-casting errors. An awkward and of-course-too-short-to-make-the-point example. No time for a 1024-lines sample:

types={
'Airport Name: MyType,
Airport Location: MyType,
[..]
'Flight Date': MyType,
'Flight Duration': MyType,
'Flight Type': MyType,
[..]
'Pilot name': MyType,
'Plane type': MyType,
'Plane buider': MyType,
[..]
}

Do the feed and, as @Skinkie put it, "_for the headers provided, in any order, this will be the matching type, but if you encounter any other column do auto_detect=true_"

**Comment by pdet** [comment]
Hi @skinkie, could you please provide an example with the exact syntax and the desired result for the new option you are proposing?

Currently, it is already possible to specify types for a subset of the columns from a CSV file while auto-detecting the remaining ones. This can be done using the `column_types` parameter, which is also aliased to `types` or `dtypes`.

There are two different ways to accomplish this:

1. You can provide the types as a struct. The struct should include the name of the column and the defined type, and it will replace them accordingly.

2. Alternatively, you can provide the types as a list, and they will be replaced in the order they are defined in the CSV file.

As an example, take the following CSV file:

```csv
a;b;c
1;bla;10
```

1. Using a struct will replace the type of a specified column name. In this example, the last column `c` will have its type set to `SMALLINT`. 
```sql
select typeof(a), typeof(b), typeof(c) FROM read_csv_auto('rc.csv',types={'c': 'SMALLINT'})
----
[('BIGINT', 'VARCHAR', 'SMALLINT')]
```

2. When using a list, types will be replaced in appearance order. In this case, only the first column `a`
```sql
select typeof(a), typeof(b), typeof(c) FROM read_csv_auto('rc.csv',types=['SMALLINT']
----
[('SMALLINT', 'VARCHAR', 'BIGINT')]
```

**Reply by FrancoisLepoutre** [reply]
Hi Pedro,

I have not tested the Duck for quite a while. That was in 2021... 

It looks like the package is progressing fast and already doing a lot better than two years ago when it comes to ingesting csv-s.  Eagerly looking forward to getting in touch again with our great _palmiped_! 

Thanks for the clarification :-)

**Reply by skinkie** [reply]
@pdet thanks for the 'types' enlightenment. That goes into the direction what I want. Sadly not all the way.

```
CREATE TABLE routes AS SELECT * FROM read_csv('/tmp/eu/routes.txt', delim=',', header=true, types={'route_type': smallint, 'bikes_allowed': boolean, 'wheelchair_accessible': boolean}, auto_detect=true);
Error: Binder Error: COLUMN_TYPES error: Columns with names: "wheelchair_accessible" do not exist in the CSV File
```

Considering the output above.
1. I don't want to be punished for providing a type which is not part of the original CSV file.
2. I would love the ability to provide missing columns, which must be created, if not already exists, according to the type provided.
3. Considering the comment of @FrancoisLepoutre above, the ability to directly alias a column name to something else might open a new can of worms (for example the ability to directly upper or lower case the input).

For point 2. A concrete example would be that `network_id` is missing as column in the source file. I am obviously not willing to change all my queries to remove `network_id` from its select part. I am looking for something like `mandatory=['route_id', 'route_type', 'network_id']` I would find it reasonable that the mandatory statement be a list where the types are being provided in `types` or it being a dict. An alias might look like `alias={'routeid': 'route_id', 'ROUTE_ID': 'route_id}`. Going into this direction, I would consider removing columns via the SELECT part, using COALESCE is not working for the aliasing or missing fields, considering that the columns do not exist in the first place.

For the purists: both `bikes_allowed` and `wheelchair_accessible` are not part of the standard.

**Reply by pdet** [reply]
For clarification, could you please share the top 3 lines of your CSV file, including the header?

For `1.`, when you mention providing the type for a column that does not exist in the CSV file, Right?

For `2.`,  are you suggesting that if a column doesn't exist, it should be created with the proposed type and column name (and I assume initializing all its values to NULL)?

Regarding `3.`In DuckDB, column names from a table (note that internally scanning a CSV file produces an in-memory table representation in DuckDB) must be unique in name, and they are case-insensitive. Therefore, I don't see how an `alias` option would fit in terms of query execution. Or do you want to alias, to use the aliased-columns as part of other options of the CSV file?

**Reply by skinkie** [reply]
```
agency_id,route_id,route_short_name,route_long_name,route_type,route_desc,route_url,route_color,route_text_color,bikes_allowed,route_sort_order
FLIXBUS-eu,N862,FlixBus N862,Prague - Pula,3,,,73D700,FFFFFF,,
FLIXBUS-eu,858,FlixBus 858,Prague - Chomutov - Most,3,,,73D700,FFFFFF,,
```

1. Right. But more broadly speaking, that does not exists in this particular delivery, in the next one it may exist.
2. Exactly. That would be nice, and would allow any defined queries to run without modification.
3. I would like to see that a CSV-column with `routeid` is renamed `route_id`, then the type of route_id is applied.
